In [3]:
import pandas as pd
import numpy as np

In [4]:
reviews = pd.read_csv('./Dataset/review.csv')

In [5]:
business = pd.read_csv('./Dataset/business.csv', dtype={'city': 'category', 
                                                                                                                        'state': 'category',
                                                                                                                        'postal_code': 'category'})

In [6]:
reviews.head()

review_id                 user_id             business_id  \
0  KU_O5udG6zpxOg-VcAEodg  mh_-eMZ6K5RLWhZyISBhwA  XQfwVwDr-v0ZS3_CbbE5Xw   
1  Sx8TMOWLNuJBWer-0pcmoA  bcjbaE6dDog4jkNY91ncLQ  e4Vwtrqf-wpJfwesgvdgxQ   
2  lUUhg8ltDsUZ9h0xnwY4Dg  RreNy--tOmXMl1en0wiBOg  cPepkJeRMtHapc_b2Oe_dw   
3  -P5E9BYUaK7s3PwBF5oAyg  Jha0USGDMefGFRLik_xFQg  bMratNjTG5ZFEA6hVyr-xQ   
4  YbMyvlDA2W3Py5lTz8VK-A  4hBhtCSgoxkrFgHa4YAD-w  bbEXAEFr4RYHLlZ-HFssTA   

   stars  useful  funny  cool  \
0    3.0     0.0    0.0   0.0   
1    4.0     1.0    0.0   1.0   
2    4.0     1.0    0.0   1.0   
3    5.0     0.0    0.0   0.0   
4    5.0     0.0    0.0   0.0   

                                                text                 date  
0  If you decide to eat here, just be aware it is...  2018-07-07 22:09:11  
1  Cute interior and owner (?) gave us tour of up...  2017-01-14 20:54:15  
2  I was really between 3 and 4 stars for this on...  2018-07-17 03:30:07  
3  First time there and it was excellent!!! It fe...  2017-02-19 13:32:05  
4  Great burgers,fries and salad!  Burgers have a...  2017-01-02 03:17:34

# P7. Teniendo en cuenta las reviews que reciben los locales, devolver una del local que haya recibido más votos del tipo divertido en sus reviews.

In [7]:
funny_reviews_per_business = reviews.groupby('business_id').agg({'funny': 'count'})
funny_reviews_per_business

funny
business_id                  
---kPU91CF4Lq2-WlRu9Lw     24
--0iUa4sNDFiZFrAdIWhZQ      2
--30_8IhuyMHbSOcNWd6DQ      4
--7PUidqRWpRSpXebiyxTg      4
--7jw19RH9JKXgFohspgQw      5
...                       ...
zzjFdJwXuxBOGe9JeY_EMw     32
zznJox6-nmXlGYNWgTDwQQ     21
zznZqH9CiAznbkV6fXyHWA     12
zztOG2cKm87I6Iw_tleZsQ      5
zzw66H6hVjXQEt0Js3Mo4A      4

[135542 rows x 1 columns]

In [8]:
business_with_max_funny_reviews = funny_reviews_per_business['funny'].idxmax()
business_with_max_funny_reviews

'ac1AeYqs8Z4_e2X5M3if2A'

## Respuesta P7

In [9]:
reviews.loc[reviews['business_id'] == business_with_max_funny_reviews].head(1)['text']

2675581    Disappointing experience. ITEMS ON MENU DO NOT...
Name: text, dtype: object

# P15. ¿Cuál es el usuario más quejumbroso por estado? Consideramos quejumbroso a un usuario que tiene más de 5 reviews y son todas de 2 estrellas o menos.

In [10]:
reviews_per_user = reviews.groupby('user_id').agg({'review_id': 'count'})
reviews_per_user.reset_index(inplace=True)
reviews_per_user.head()

user_id  review_id
0  ---1lKK3aKOuomHnwAkAow          1
1  ---2PmXbF47D870stH1jqA          3
2  ---UgP94gokyCDuB5zUssA          6
3  ---fa6ZK37T9NjkGKI4oSg          1
4  ---r61b7EpVPkb4UVme5tA          1

In [11]:
reviews_per_user.rename(columns = {'review_id':'reviews_per_user'}, inplace = True)
reviews_per_user.head()

user_id  reviews_per_user
0  ---1lKK3aKOuomHnwAkAow                 1
1  ---2PmXbF47D870stH1jqA                 3
2  ---UgP94gokyCDuB5zUssA                 6
3  ---fa6ZK37T9NjkGKI4oSg                 1
4  ---r61b7EpVPkb4UVme5tA                 1

In [12]:
reviews_with_reviews_per_user = reviews.merge(reviews_per_user, how='left', on='user_id')
reviews_with_reviews_per_user.head()

review_id                 user_id             business_id  \
0  KU_O5udG6zpxOg-VcAEodg  mh_-eMZ6K5RLWhZyISBhwA  XQfwVwDr-v0ZS3_CbbE5Xw   
1  Sx8TMOWLNuJBWer-0pcmoA  bcjbaE6dDog4jkNY91ncLQ  e4Vwtrqf-wpJfwesgvdgxQ   
2  lUUhg8ltDsUZ9h0xnwY4Dg  RreNy--tOmXMl1en0wiBOg  cPepkJeRMtHapc_b2Oe_dw   
3  -P5E9BYUaK7s3PwBF5oAyg  Jha0USGDMefGFRLik_xFQg  bMratNjTG5ZFEA6hVyr-xQ   
4  YbMyvlDA2W3Py5lTz8VK-A  4hBhtCSgoxkrFgHa4YAD-w  bbEXAEFr4RYHLlZ-HFssTA   

   stars  useful  funny  cool  \
0    3.0     0.0    0.0   0.0   
1    4.0     1.0    0.0   1.0   
2    4.0     1.0    0.0   1.0   
3    5.0     0.0    0.0   0.0   
4    5.0     0.0    0.0   0.0   

                                                text                 date  \
0  If you decide to eat here, just be aware it is...  2018-07-07 22:09:11   
1  Cute interior and owner (?) gave us tour of up...  2017-01-14 20:54:15   
2  I was really between 3 and 4 stars for this on...  2018-07-17 03:30:07   
3  First time there and it was excellent!!! It fe...  2017-02-19 13:32:05   
4  Great burgers,fries and salad!  Burgers have a...  2017-01-02 03:17:34   

   reviews_per_user  
0                28  
1                66  
2               107  
3                77  
4                22

In [13]:
reviews_by_reviewers_with_more_than_5 = reviews_with_reviews_per_user[reviews_with_reviews_per_user['reviews_per_user'] > 5]

In [14]:
reviews_by_reviewers_with_more_than_5.head()

review_id                 user_id             business_id  \
0  KU_O5udG6zpxOg-VcAEodg  mh_-eMZ6K5RLWhZyISBhwA  XQfwVwDr-v0ZS3_CbbE5Xw   
1  Sx8TMOWLNuJBWer-0pcmoA  bcjbaE6dDog4jkNY91ncLQ  e4Vwtrqf-wpJfwesgvdgxQ   
2  lUUhg8ltDsUZ9h0xnwY4Dg  RreNy--tOmXMl1en0wiBOg  cPepkJeRMtHapc_b2Oe_dw   
3  -P5E9BYUaK7s3PwBF5oAyg  Jha0USGDMefGFRLik_xFQg  bMratNjTG5ZFEA6hVyr-xQ   
4  YbMyvlDA2W3Py5lTz8VK-A  4hBhtCSgoxkrFgHa4YAD-w  bbEXAEFr4RYHLlZ-HFssTA   

   stars  useful  funny  cool  \
0    3.0     0.0    0.0   0.0   
1    4.0     1.0    0.0   1.0   
2    4.0     1.0    0.0   1.0   
3    5.0     0.0    0.0   0.0   
4    5.0     0.0    0.0   0.0   

                                                text                 date  \
0  If you decide to eat here, just be aware it is...  2018-07-07 22:09:11   
1  Cute interior and owner (?) gave us tour of up...  2017-01-14 20:54:15   
2  I was really between 3 and 4 stars for this on...  2018-07-17 03:30:07   
3  First time there and it was excellent!!! It fe...  2017-02-19 13:32:05   
4  Great burgers,fries and salad!  Burgers have a...  2017-01-02 03:17:34   

   reviews_per_user  
0                28  
1                66  
2               107  
3                77  
4                22

Ahora que tengo a todos los que tienen más de 5 reviews necesito a los más quejumbrosos.

In [15]:
good_reviewers = reviews_by_reviewers_with_more_than_5[reviews_by_reviewers_with_more_than_5['stars'] >= 3]
good_reviewers.head()

review_id                 user_id             business_id  \
0  KU_O5udG6zpxOg-VcAEodg  mh_-eMZ6K5RLWhZyISBhwA  XQfwVwDr-v0ZS3_CbbE5Xw   
1  Sx8TMOWLNuJBWer-0pcmoA  bcjbaE6dDog4jkNY91ncLQ  e4Vwtrqf-wpJfwesgvdgxQ   
2  lUUhg8ltDsUZ9h0xnwY4Dg  RreNy--tOmXMl1en0wiBOg  cPepkJeRMtHapc_b2Oe_dw   
3  -P5E9BYUaK7s3PwBF5oAyg  Jha0USGDMefGFRLik_xFQg  bMratNjTG5ZFEA6hVyr-xQ   
4  YbMyvlDA2W3Py5lTz8VK-A  4hBhtCSgoxkrFgHa4YAD-w  bbEXAEFr4RYHLlZ-HFssTA   

   stars  useful  funny  cool  \
0    3.0     0.0    0.0   0.0   
1    4.0     1.0    0.0   1.0   
2    4.0     1.0    0.0   1.0   
3    5.0     0.0    0.0   0.0   
4    5.0     0.0    0.0   0.0   

                                                text                 date  \
0  If you decide to eat here, just be aware it is...  2018-07-07 22:09:11   
1  Cute interior and owner (?) gave us tour of up...  2017-01-14 20:54:15   
2  I was really between 3 and 4 stars for this on...  2018-07-17 03:30:07   
3  First time there and it was excellent!!! It fe...  2017-02-19 13:32:05   
4  Great burgers,fries and salad!  Burgers have a...  2017-01-02 03:17:34   

   reviews_per_user  
0                28  
1                66  
2               107  
3                77  
4                22

In [16]:
good_reviewers_ids = good_reviewers['user_id'].unique()

In [17]:
bad_reviewers_reviews = reviews_by_reviewers_with_more_than_5[~reviews_by_reviewers_with_more_than_5['user_id'].isin(good_reviewers_ids)]
bad_reviewers_reviews.head()

review_id                 user_id             business_id  \
1692  BY77fmsL9bj3e7paMdJGuw  bg7Wnf_EB3Rh8Asr8BApIg  xCW-s1hBP23-HkyqW-kngA   
1911  Gy66c2CKhEMFC231Gqykjw  APB7OgxLasHDj4MoRu8XIQ  Zdw5rbtAWguwcT9h_X4xRA   
2397  rcuvpXb75km8m0YixC6Wdg  0x8EK4LCGMCP8Zrlvelr-g  eS4billU3Yd9hoVUFfZFbA   
2716  Plak841p816V5yaTcbU1Jg  jJZ21Ur5K02886Z6Eq_SnQ  vPK_MN51evy0007W8NeJ6w   
4298  noIwwrpSYv0YPEpQsQk6Tg  Q0GISGD6Vy-u3lTFuHU_5Q  bJsOFjNgVl7wcy-by6z8KQ   

      stars  useful  funny  cool  \
1692    1.0     0.0    0.0   0.0   
1911    2.0     0.0    0.0   0.0   
2397    2.0     0.0    0.0   0.0   
2716    1.0     0.0    0.0   0.0   
4298    1.0     0.0    0.0   0.0   

                                                   text                 date  \
1692  If i could give this place a zero stars I woul...  2017-03-04 01:05:40   
1911  Not pet friendly which is quite odd since it's...  2018-07-08 01:44:25   
2397  Must have changed management. I go about twice...  2017-06-06 21:39:32   
2716  Don't bother to place an order in person.  Wit...  2017-08-29 01:13:00   
4298  They send many jobs over seas to phillipines a...  2018-06-09 20:36:57   

      reviews_per_user  
1692                13  
1911                 9  
2397                 6  
2716                 7  
4298                22

In [18]:
business.head()

business_id                      name  \
0  Pns2l4eNsfO8kk83dixA6A  Abby Rappoport, LAC, CMQ   
1  mpf3x-BjTdTEA3yCZrAYPw             The UPS Store   
2  tUFrWirKiKi_TAnsVWINQQ                    Target   
3  MTSW4McQd7CbVtyjqoe9mw        St Honore Pastries   
4  mWMc6_wTdE0EUBKIGXDVfA  Perkiomen Valley Brewery   

                           address           city state postal_code  \
0           1616 Chapala St, Ste 2  Santa Barbara    CA       93101   
1  87 Grasso Plaza Shopping Center         Affton    MO       63123   
2             5255 E Broadway Blvd         Tucson    AZ       85711   
3                      935 Race St   Philadelphia    PA       19107   
4                    101 Walnut St     Green Lane    PA       18054   

    latitude   longitude  stars  review_count  is_open  \
0  34.426679 -119.711197    5.0           0.0        0   
1  38.551126  -90.335695    3.0           8.0        1   
2  32.223236 -110.880452    3.5          11.0        0   
3  39.955505  -75.155564    4.0          37.0        1   
4  40.338183  -75.471659    4.5          13.0        1   

                                          attributes  \
0                      {'ByAppointmentOnly': 'True'}   
1             {'BusinessAcceptsCreditCards': 'True'}   
2  {'BikeParking': 'True', 'BusinessAcceptsCredit...   
3  {'RestaurantsDelivery': 'False', 'OutdoorSeati...   
4  {'BusinessAcceptsCreditCards': 'True', 'Wheelc...   

                                          categories  \
0  Doctors, Traditional Chinese Medicine, Naturop...   
1  Shipping Centers, Local Services, Notaries, Ma...   
2  Department Stores, Shopping, Fashion, Home & G...   
3  Restaurants, Food, Bubble Tea, Coffee & Tea, B...   
4                          Brewpubs, Breweries, Food   

                                               hours  
0                                                NaN  
1  {'Monday': '0:0-0:0', 'Tuesday': '8:0-18:30', ...  
2  {'Monday': '8:0-22:0', 'Tuesday': '8:0-22:0', ...  
3  {'Monday': '7:0-20:0', 'Tuesday': '7:0-20:0', ...  
4  {'Wednesday': '14:0-22:0', 'Thursday': '16:0-2...

In [19]:
business.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150346 entries, 0 to 150345
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype   
---  ------        --------------   -----   
 0   business_id   150346 non-null  object  
 1   name          150346 non-null  object  
 2   address       145219 non-null  object  
 3   city          150346 non-null  category
 4   state         150346 non-null  category
 5   postal_code   150273 non-null  category
 6   latitude      150346 non-null  float64 
 7   longitude     150346 non-null  float64 
 8   stars         150346 non-null  float64 
 9   review_count  150346 non-null  float64 
 10  is_open       150346 non-null  int64   
 11  attributes    136602 non-null  object  
 12  categories    150243 non-null  object  
 13  hours         127123 non-null  object  
dtypes: category(3), float64(4), int64(1), object(6)
memory usage: 13.5+ MB


In [20]:
bad_reviewers_reviews_with_business_data = bad_reviewers_reviews.merge(business, how='left', on='business_id')
bad_reviewers_reviews_with_business_data.head()

review_id                 user_id             business_id  \
0  BY77fmsL9bj3e7paMdJGuw  bg7Wnf_EB3Rh8Asr8BApIg  xCW-s1hBP23-HkyqW-kngA   
1  Gy66c2CKhEMFC231Gqykjw  APB7OgxLasHDj4MoRu8XIQ  Zdw5rbtAWguwcT9h_X4xRA   
2  rcuvpXb75km8m0YixC6Wdg  0x8EK4LCGMCP8Zrlvelr-g  eS4billU3Yd9hoVUFfZFbA   
3  Plak841p816V5yaTcbU1Jg  jJZ21Ur5K02886Z6Eq_SnQ  vPK_MN51evy0007W8NeJ6w   
4  noIwwrpSYv0YPEpQsQk6Tg  Q0GISGD6Vy-u3lTFuHU_5Q  bJsOFjNgVl7wcy-by6z8KQ   

   stars_x  useful  funny  cool  \
0      1.0     0.0    0.0   0.0   
1      2.0     0.0    0.0   0.0   
2      2.0     0.0    0.0   0.0   
3      1.0     0.0    0.0   0.0   
4      1.0     0.0    0.0   0.0   

                                                text                 date  \
0  If i could give this place a zero stars I woul...  2017-03-04 01:05:40   
1  Not pet friendly which is quite odd since it's...  2018-07-08 01:44:25   
2  Must have changed management. I go about twice...  2017-06-06 21:39:32   
3  Don't bother to place an order in person.  Wit...  2017-08-29 01:13:00   
4  They send many jobs over seas to phillipines a...  2018-06-09 20:36:57   

   reviews_per_user  ... state postal_code   latitude  longitude stars_y  \
0                13  ...    FL       33569  27.881078 -82.328286     4.0   
1                 9  ...    NJ       08055  39.919560 -74.851228     3.5   
2                 6  ...    FL       33761  28.033825 -82.739191     4.0   
3                 7  ...    FL       33596  27.894241 -82.257515     3.0   
4                22  ...    IL       62025  38.781757 -89.951784     1.5   

   review_count  is_open                                         attributes  \
0          93.0        1  {'RestaurantsPriceRange2': '2', 'BusinessAccep...   
1          83.0        1  {'RestaurantsPriceRange2': '2', 'BusinessParki...   
2          18.0        0  {'NoiseLevel': "u'average'", 'RestaurantsReser...   
3          12.0        1  {'BusinessAcceptsCreditCards': 'True', 'Restau...   
4          18.0        1  {'BusinessParking': "{'garage': False, 'street...   

                                          categories  \
0         Italian, Internet Cafes, Restaurants, Food   
1             Food, Specialty Food, Fruits & Veggies   
2  Hot Dogs, Sandwiches, American (New), Salad, F...   
3      Chicken Wings, Pizza, Sandwiches, Restaurants   
4  Home Services, Television Service Providers, L...   

                                               hours  
0  {'Monday': '12:0-22:0', 'Tuesday': '12:0-22:0'...  
1  {'Monday': '8:0-19:0', 'Tuesday': '8:0-19:0', ...  
2  {'Monday': '11:0-21:0', 'Tuesday': '11:0-21:0'...  
3  {'Monday': '10:30-0:0', 'Tuesday': '10:30-0:0'...  
4  {'Monday': '9:0-21:0', 'Tuesday': '9:0-21:0', ...  

[5 rows x 23 columns]

In [21]:
bad_reviewers_reviews_with_business_data[['review_id', 'user_id', 'stars_x', 'state', 'reviews_per_user']]

review_id                 user_id  stars_x state  \
0     BY77fmsL9bj3e7paMdJGuw  bg7Wnf_EB3Rh8Asr8BApIg      1.0    FL   
1     Gy66c2CKhEMFC231Gqykjw  APB7OgxLasHDj4MoRu8XIQ      2.0    NJ   
2     rcuvpXb75km8m0YixC6Wdg  0x8EK4LCGMCP8Zrlvelr-g      2.0    FL   
3     Plak841p816V5yaTcbU1Jg  jJZ21Ur5K02886Z6Eq_SnQ      1.0    FL   
4     noIwwrpSYv0YPEpQsQk6Tg  Q0GISGD6Vy-u3lTFuHU_5Q      1.0    IL   
...                      ...                     ...      ...   ...   
7457  -eRd4u_7PSMTuA8owu9nzw  4RpBsI1PNE32O4ryKVtBeQ      2.0    PA   
7458  F_6xs821Fgx42h8M7r9zgQ  XaNmvT-8hKkrVg7xwbCwRw      2.0    PA   
7459  Te4vW9hwWK-NMU_3QfbG2Q  EGO6fKSG5qtZHfX8p3nPFQ      1.0    NV   
7460  oJm1t0gFfFwV3qi2k3TVUA  u7waQI6HNCTJ6aEIxt1NMQ      1.0    IN   
7461  GunwjtPlDcf8HyRycpt-Pg  GmzkFUeiyHcxHb0lB-HaHw      1.0    PA   

      reviews_per_user  
0                   13  
1                    9  
2                    6  
3                    7  
4                   22  
...                ...  
7457                 6  
7458                 7  
7459                 7  
7460                 6  
7461                 6  

[7462 rows x 5 columns]

In [22]:
reviews_by_bad_reviewers_with_state = bad_reviewers_reviews_with_business_data[['review_id', 'user_id', 'stars_x', 'state', 'reviews_per_user']].rename(columns={
    'stars_x': 'stars'
})
reviews_by_bad_reviewers_with_state.head()

review_id                 user_id  stars state  \
0  BY77fmsL9bj3e7paMdJGuw  bg7Wnf_EB3Rh8Asr8BApIg    1.0    FL   
1  Gy66c2CKhEMFC231Gqykjw  APB7OgxLasHDj4MoRu8XIQ    2.0    NJ   
2  rcuvpXb75km8m0YixC6Wdg  0x8EK4LCGMCP8Zrlvelr-g    2.0    FL   
3  Plak841p816V5yaTcbU1Jg  jJZ21Ur5K02886Z6Eq_SnQ    1.0    FL   
4  noIwwrpSYv0YPEpQsQk6Tg  Q0GISGD6Vy-u3lTFuHU_5Q    1.0    IL   

   reviews_per_user  
0                13  
1                 9  
2                 6  
3                 7  
4                22

In [23]:
bad_reviewers_per_state = reviews_by_bad_reviewers_with_state.groupby(['state', 'user_id']).agg({'review_id': 'count'})
bad_reviewers_per_state

review_id
state user_id                          
AB    --FEYzujiiMt7r5QCvX8YQ          0
      --rTlRNPla6azelpN66Y6A          0
      -5Yx5X3OxosNU5gojBug8A          0
      -6SIeb756UFhLbVLQlVhiQ          0
      -CBB60Pd0_hWkvk3dc8otw          0
...                                 ...
VT    zb723R2g4erHBoPh8t5I-w          0
      zbz4HWrL424kKXDNF24y9A          0
      zlCb9yTLLqvl_tHrQmwTEA          0
      zrGkh3ppL13VjoaiB66p6g          0
      zwP7nyqroYJ_XweRkrRjng          0

[26541 rows x 1 columns]

In [24]:
reviews_by_bad_reviewers_with_state.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7462 entries, 0 to 7461
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   review_id         7462 non-null   object  
 1   user_id           7462 non-null   object  
 2   stars             7462 non-null   float64 
 3   state             7462 non-null   category
 4   reviews_per_user  7462 non-null   int64   
dtypes: category(1), float64(1), int64(1), object(2)
memory usage: 300.0+ KB


In [25]:
max_bad_reviewer_per_state = reviews_by_bad_reviewers_with_state.groupby('state').max()
max_bad_reviewer_per_state

review_id                 user_id  stars  reviews_per_user
state                                                                         
AB     z67JmINnH-mdjXDK1izdpg  qZN5erVChz14pP1CMJFsoA    2.0               9.0
AZ     zzENKTBTJrJYt4iEC3-NNg  yl-gm-Q1GEur4nMJaq3xng    2.0              28.0
CA     zz4A2uioJAmKEcEyLm-SXA  xIJsXMZoUBwfrS92ahGqcA    2.0              10.0
CO                       None                    None    NaN               NaN
DE     z0sMVqa8g667DGH7dk-R4A  zb723R2g4erHBoPh8t5I-w    2.0              23.0
FL     zx0_dd2u-ioZ66lUsRt1IA  zX6viSOn2Oooob4jFkV2bA    2.0              32.0
HI                       None                    None    NaN               NaN
ID     zvTijXRgOGbEJzQZApYE6g  yLwVAJDrKMGwSpRy1kTXMA    2.0              14.0
IL     yu4lOR24jXjZgd3dsvUvtQ  zYHLi9zaxHmI6RQVxOo7hg    2.0              32.0
IN     zmGkdtXL1T0BIYm15rMsIw  yDJcA1RKGFMr7rblL3LcvQ    2.0              13.0
LA     zyGAJ7O5sES5Igakce0K2Q  y_pa-o1iTY8cd1YGoI8IYg    2.0              16.0
MI                       None                    None    NaN               NaN
MO     zpY6cHSb5UJi6Y68-6fJuw  yq-NVs3XR2FSumZjGwtyVQ    2.0              32.0
MT                       None                    None    NaN               NaN
NC                       None                    None    NaN               NaN
NJ     zx8DPhwgJIfIMTAh_sKYIg  zwP7nyqroYJ_XweRkrRjng    2.0              15.0
NV     zzwhy9kNTWirm4ChNbWWtA  zlCb9yTLLqvl_tHrQmwTEA    2.0              22.0
PA     zzzeRDxTrb_cPKVq25WYmQ  zwP7nyqroYJ_XweRkrRjng    2.0              24.0
TN     zyG2IRZDRwTfDDhGRxyqpw  zrGkh3ppL13VjoaiB66p6g    2.0              21.0
TX                       None                    None    NaN               NaN
UT                       None                    None    NaN               NaN
WA                       None                    None    NaN               NaN
MA                       None                    None    NaN               NaN
SD                       None                    None    NaN               NaN
XMS                      None                    None    NaN               NaN
VI                       None                    None    NaN               NaN
VT                       None                    None    NaN               NaN

## ¿Por qué hay estados en None si yo armé el dataset en base a reviewers que cumplían mis condiciones?

Esto es porque 'state' está definida como una columna de tipo category, entonces groupby está mostrando inclusive categorías que no utilizamos en nuestra tabla. 
Para probar esto reviso, por ejemplo uno de los estados que me dio None.

In [26]:
reviews_by_bad_reviewers_with_state[reviews_by_bad_reviewers_with_state['state'] == 'HI']

Empty DataFrame
Columns: [review_id, user_id, stars, state, reviews_per_user]
Index: []

## Respuesta

In [27]:
max_bad_reviewer_per_state.fillna('---')

review_id                 user_id stars reviews_per_user
state                                                                       
AB     z67JmINnH-mdjXDK1izdpg  qZN5erVChz14pP1CMJFsoA   2.0              9.0
AZ     zzENKTBTJrJYt4iEC3-NNg  yl-gm-Q1GEur4nMJaq3xng   2.0             28.0
CA     zz4A2uioJAmKEcEyLm-SXA  xIJsXMZoUBwfrS92ahGqcA   2.0             10.0
CO                        ---                     ---   ---              ---
DE     z0sMVqa8g667DGH7dk-R4A  zb723R2g4erHBoPh8t5I-w   2.0             23.0
FL     zx0_dd2u-ioZ66lUsRt1IA  zX6viSOn2Oooob4jFkV2bA   2.0             32.0
HI                        ---                     ---   ---              ---
ID     zvTijXRgOGbEJzQZApYE6g  yLwVAJDrKMGwSpRy1kTXMA   2.0             14.0
IL     yu4lOR24jXjZgd3dsvUvtQ  zYHLi9zaxHmI6RQVxOo7hg   2.0             32.0
IN     zmGkdtXL1T0BIYm15rMsIw  yDJcA1RKGFMr7rblL3LcvQ   2.0             13.0
LA     zyGAJ7O5sES5Igakce0K2Q  y_pa-o1iTY8cd1YGoI8IYg   2.0             16.0
MI                        ---                     ---   ---              ---
MO     zpY6cHSb5UJi6Y68-6fJuw  yq-NVs3XR2FSumZjGwtyVQ   2.0             32.0
MT                        ---                     ---   ---              ---
NC                        ---                     ---   ---              ---
NJ     zx8DPhwgJIfIMTAh_sKYIg  zwP7nyqroYJ_XweRkrRjng   2.0             15.0
NV     zzwhy9kNTWirm4ChNbWWtA  zlCb9yTLLqvl_tHrQmwTEA   2.0             22.0
PA     zzzeRDxTrb_cPKVq25WYmQ  zwP7nyqroYJ_XweRkrRjng   2.0             24.0
TN     zyG2IRZDRwTfDDhGRxyqpw  zrGkh3ppL13VjoaiB66p6g   2.0             21.0
TX                        ---                     ---   ---              ---
UT                        ---                     ---   ---              ---
WA                        ---                     ---   ---              ---
MA                        ---                     ---   ---              ---
SD                        ---                     ---   ---              ---
XMS                       ---                     ---   ---              ---
VI                        ---                     ---   ---              ---
VT                        ---                     ---   ---              ---

# P20. Buscar la distancia mínima que existe entre dos locales de McDonald ‘s en el estado de LA (que no comparten dirección). ¿Cuál es la distancia?¿A qué ciudad/es pertenecen? ¿Cuáles son las direcciones de los locales?

In [28]:
mcdonadls_in_la = business[(business['state'] == 'LA') & (business['name'] == 'McDonald\'s' )]
mcdonadls_in_la.head()

business_id        name                      address  \
5923   -9yzQQ0d_rcOD2CzdTNO_Q  McDonald's               4240 Louisa St   
10078  wPwxfwrSNMmqxEHMzGG1Jw  McDonald's                  2951 Hwy 90   
19709  wjUlOsBLLnxpprgI-sSiDA  McDonald's  5315 Veterans Memorial Blvd   
21693  UxPNpzKEC4FpdidHz8MCcA  McDonald's            3443 S Carrollton   
25554  sIgFetlP6tX1vYWYjkyq1w  McDonald's             301 Veterans Hwy   

              city state postal_code   latitude  longitude  stars  \
5923   New Orleans    LA       70126  30.004244 -90.036407    2.0   
10078     Avondale    LA       70094  29.909168 -90.204384    1.0   
19709     Metairie    LA       70003  30.006814 -90.200752    1.5   
21693  New Orleans    LA       70118  29.963289 -90.111832    2.5   
25554     Metairie    LA       70005  30.001218 -90.126002    1.5   

       review_count  is_open  \
5923           11.0        1   
10078          10.0        1   
19709          23.0        1   
21693          18.0        1   
25554          41.0        1   

                                              attributes  \
5923   {'HasTV': 'True', 'NoiseLevel': "u'loud'", 'Re...   
10078  {'RestaurantsGoodForGroups': 'True', 'Restaura...   
19709  {'BusinessParking': "{'garage': False, 'street...   
21693  {'GoodForKids': 'True', 'BusinessParking': "{'...   
25554  {'OutdoorSeating': 'True', 'GoodForKids': 'Tru...   

                                              categories  \
5923   Fast Food, Restaurants, Coffee & Tea, Food, Bu...   
10078  Sandwiches, Coffee & Tea, Burgers, Food, Resta...   
19709  Food, Burgers, Fast Food, Coffee & Tea, Restau...   
21693  Coffee & Tea, Food, Restaurants, Burgers, Fast...   
25554  Coffee & Tea, Restaurants, Food, Fast Food, Bu...   

                                                   hours  
5923   {'Monday': '0:0-0:0', 'Tuesday': '0:0-0:0', 'W...  
10078  {'Monday': '5:0-0:0', 'Tuesday': '5:0-0:0', 'W...  
19709  {'Monday': '0:0-0:0', 'Tuesday': '0:0-0:0', 'W...  
21693  {'Monday': '0:0-0:0', 'Tuesday': '0:0-0:0', 'W...  
25554  {'Monday': '0:0-0:0', 'Tuesday': '0:0-0:0', 'W...

In [29]:
len(mcdonadls_in_la)

42

In [30]:
mcdonadls_in_la = mcdonadls_in_la[['business_id', 'latitude', 'longitude', 'address', 'city']]

In [31]:
mcdonadls_in_la.head()

business_id   latitude  longitude  \
5923   -9yzQQ0d_rcOD2CzdTNO_Q  30.004244 -90.036407   
10078  wPwxfwrSNMmqxEHMzGG1Jw  29.909168 -90.204384   
19709  wjUlOsBLLnxpprgI-sSiDA  30.006814 -90.200752   
21693  UxPNpzKEC4FpdidHz8MCcA  29.963289 -90.111832   
25554  sIgFetlP6tX1vYWYjkyq1w  30.001218 -90.126002   

                           address         city  
5923                4240 Louisa St  New Orleans  
10078                  2951 Hwy 90     Avondale  
19709  5315 Veterans Memorial Blvd     Metairie  
21693            3443 S Carrollton  New Orleans  
25554             301 Veterans Hwy     Metairie

In [32]:

def haversine_np(lon1, lat1, lon2, lat2):
    """
    Calculate the great circle distance between two points
    on the earth (specified in decimal degrees)

    All args must be of equal length.    

    """
    lon1, lat1, lon2, lat2 = map(np.radians, [lon1, lat1, lon2, lat2])

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = np.sin(dlat/2.0)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2.0)**2

    c = 2 * np.arcsin(np.sqrt(a))
    km = 6367 * c
    return km

In [33]:
# position = 0
df = mcdonadls_in_la.merge(mcdonadls_in_la, how='cross')
df['distance'] = haversine_np(df['longitude_x'], df['latitude_x'],
                                                           df['longitude_y'], df['latitude_y'])
df.drop(df[df['distance'] == 0].index, inplace=True)
# print(df['distance'].min())
# print(df['distance'].idxmin())
df.loc[[df['distance'].idxmin()]]

business_id_x  latitude_x  longitude_x          address_x  \
706  xZxAoUnClT5wm0-QuZPFwg   29.910136   -90.057234  501 Westbank Expy   

     city_x           business_id_y  latitude_y  longitude_y  \
706  Gretna  zZKnptPeSPSkza6SKq00HQ    29.91461   -90.046379   

            address_y  city_y  distance  
706  1212 Stumpf Blvd  Gretna  1.157798

In [34]:
df.sample(10)

business_id_x  latitude_x  longitude_x  \
154   UxPNpzKEC4FpdidHz8MCcA   29.963289   -90.111832   
781   XU8MGYX9oKQ0bl4r8IMqfQ   29.973762   -90.151390   
1606  I1LP0Vg55lwufnoNREE1KA   29.941588   -90.095297   
1396  d9i5hq9hIjmp7RN2hsOTzQ   29.950407   -89.959153   
1210  499j2YBfJuZLgNOk7SzFCA   29.879620   -90.098122   
410   0uAdNV5KUFOgbg42MhDwWQ   29.956305   -90.217029   
907   RMZOOXWt6S0FkzjsuwjP0g   29.953599   -89.978079   
432   1JXJbmiVaLK2Wc2-reqQ5w   29.933581   -90.007743   
1172  xbgzX-_EpV6yH-6nf2LCNw   29.926163   -90.022580   
337   JpH5dhrs9nB-DEoX2dbZwQ   29.961776   -90.183633   

                     address_x       city_x           business_id_y  \
154          3443 S Carrollton  New Orleans  499j2YBfJuZLgNOk7SzFCA   
781            2700 Airline Dr     Metairie  zdJCe6BLDB03FBC4U_9bYw   
1606          2856 S Claiborne  New Orleans  1JXJbmiVaLK2Wc2-reqQ5w   
1396             3401 Paris Rd    Chalmette  1JXJbmiVaLK2Wc2-reqQ5w   
1210       1681 Barataria Blvd      Marrero  zZKnptPeSPSkza6SKq00HQ   
410         9601 Jefferson Hwy  River Ridge  MoSyhSKj62gqBJyR_e5kIA   
907         8319 W Judge Perez    Chalmette  zdJCe6BLDB03FBC4U_9bYw   
432     4589 General Meyer Ave  New Orleans  pvGBHwMceiP6c6L41aL2tw   
1172  3250 General Degaulle Dr  New Orleans  I1LP0Vg55lwufnoNREE1KA   
337      1151 S Clearview Pkwy    Jefferson  wPwxfwrSNMmqxEHMzGG1Jw   

      latitude_y  longitude_y               address_y        city_y   distance  
154    29.879620   -90.098122     1681 Barataria Blvd       Marrero   9.390984  
781    30.038396   -89.975671   10001 I 10 Service Rd   New Orleans  18.371834  
1606   29.933581   -90.007743  4589 General Meyer Ave   New Orleans   8.478050  
1396   29.933581   -90.007743  4589 General Meyer Ave   New Orleans   5.038717  
1210   29.914610   -90.046379        1212 Stumpf Blvd        Gretna   6.321924  
410    29.969473   -90.049637   2609 Saint Claude Ave   New Orleans  16.181646  
907    30.038396   -89.975671   10001 I 10 Service Rd   New Orleans   9.425926  
432    29.862522   -89.996876             7862 Hwy 23  Belle Chasse   7.965586  
1172   29.941588   -90.095297        2856 S Claiborne   New Orleans   7.209504  
337    29.909168   -90.204384             2951 Hwy 90      Avondale   6.178163

## Respuesta
La distancia mas corta es 1.15 km y las direcciones son: 501 Westbank Expy y 1212 Stumpf Blvd, ambos ubicados en la ciudad de Gretna.